File: Algorithm/merge_sort.cpp
#include <iostream>
using namespace std;

void Merge(int arr[],int l,int m,int r)
{
    int i=l;
    int j=m+1;
    int k=l;
    int temp[7];

    while(i<=m && j<=r)
    {
        if(arr[i] <=arr[j])
        {
            temp[k]=arr[i];
            i++;
            k++;
        }
        else
        {
            temp[k]=arr[j];
            j++;
            k++;
        }
    }
    while(i<=m)
    {
        temp[k]=arr[i];
        i++;
        k++;
    }
    while(j<=r)
    {
        temp[k]=arr[j];
        j++;
        k++;
    }
    for(int p=l; p<=r;p++)
    {
        arr[p]=temp[p];
    }
}
void MergeSort(int arr[],int l,int r)
{
    int m= (l+r)/2;
    if(l<r)
    {
        MergeSort(arr,l,m);
        MergeSort(arr,m+1,r);
        Merge(arr,l,m,r);
    }

}
int main()
{
    int arr[]={9,8,7,6,5,4,3};
    MergeSort(arr,0,6);
    for(int i=0;i<7;i++)
    {
        cout<<arr[i]<<" ";
    }
}

File: merge_sort.java
public class merge_sort {

    public static void mergeSort(int arr[],int l, int r){
        if(l<r){
            int mid=(l+r)/2;
            mergeSort(arr,l,mid);
            mergeSort(arr,mid+1,r);

            merge(arr,l,r,mid);
        }
    }
    public static void merge(int arr[],int l, int r, int mid){

        //Array Size
        int n1=mid-l+1;
        int n2=r-mid;
        int arr1[]=new int[n1];
        int arr2[]=new int[n2];

        //Copying elements
        for(int a=0;a<n1;a++){
            arr1[a]=arr[l+a];
        }
        for(int b=0;b<n2;b++){
            arr2[b]=arr[mid+1+b];
        }

        //Checking Conditions
        int i=0;
        int j=0;
        int k=l;

        while(i <n1 && j< n2){
            if(arr1[i] <= arr2[j]){
                arr[k]=arr1[i];
                i++;
            }
            else {
                arr[k]=arr2[j];
                j++;
            }
            k++;
        }

        //Adding remaining elements for both arrays
        while(i < n1){
            arr[k]=arr1[i];
            i++;
        }
        while(j < n2){
            arr[k]=arr2[j];
            j++;
        }

    }
    public static void main(String[] args) {
        int arr[]={6,3,9,5,2,8};
        mergeSort(arr,0,arr.length-1);
        for(int i=0;i<arr.length;i++){
            System.out.println(arr[i]);
        }
    }
}
